storage: provide http-based elasticsearch client

sethp-jive opened the pull request:
Taking advantage of & expanding the previous commit's shim, this chnage
introduces an elasticsearch client that operates over the HTTP(S)
protocol. This commit's functional changes are minimal, limited to
restricting flushes on a few operations to the desired indices
instead of globally.

sethp-jive committed:
storage: provide http-based elasticsearch client

Taking advantage of & expanding the previous commit's shim, this chnage
introduces an elasticsearch client that operates over the HTTP(S)
protocol. This commit's functional changes are minimal, limited to
restricting flushes on a few operations to the desired indices
instead of globally.

sethp-jive commented on zipkin-storage/elasticsearch/src/main/java/zipkin/storage/elasticsearch/InternalElasticsearchClient.java:
```diff
@@ -48,7 +48,7 @@
 
   protected interface Factory {
-    InternalElasticsearchClient create();
+    InternalElasticsearchClient create(String allIndices);
```
I had to bring this back to support the index-query-rewriting thing I'm doing in the RestClient (see the javadoc attached to `MAX_INDICES`). It's definitely ugly, but IMO less so than making the client check the # of indices and do the query rewriting themselves.

codefromthecrypt commented on zipkin-storage/elasticsearch-jest/src/test/java/zipkin/storage/elasticsearch/jest/ElasticsearchJestDependencyWriter.java:
```diff
@@ -0,0 +1,47 @@
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class ElasticsearchJestDependencyWriter {
+  public static void writeDependencyLinks(InternalElasticsearchClient genericClient,
```
I'd prefer this moved to be inside the class of its only caller

codefromthecrypt commented on zipkin-storage/elasticsearch-jest/src/test/java/zipkin/storage/elasticsearch/jest/ElasticsearchJestTestGraph.java:
```diff
@@ -0,0 +1,40 @@
+                  .index("test_zipkin_rest").build();
+          CheckResult check = result.check();
+          if (check.ok) return result;
+          throw ex = new AssumptionViolatedException("cluster was not healthy", check.exception);
```
check.exception.getMessage() instead of "cluster was not healthy" as the failure could be unrelated to the cluster

sethp-jive replied:
Yup, good catch – I've fixed it up to look like the others in #1305

codefromthecrypt commented on zipkin-storage/elasticsearch/src/main/java/zipkin/storage/elasticsearch/NativeClient.java:
```diff
@@ -149,9 +149,9 @@
 
   final TransportClient client;
-  final boolean flushOnWrites;
+  private final boolean flushOnWrites;
```
not required to change, but just so you know why I didn't mark these private. This is a package protected final internal class anyway, so we have no contract to uphold wrt changing thing. each private thing has potential to bloat class sizes and mixed visibility is distracting.

If you really want these things to be invisible to the package, a quicker and less bloaty way to do it is just to make the type private. At that point, you can't reference the fields and constructors anyway, so default visibility is fine. You'll notice this practice used in grpc I think..

sethp-jive replied:
I changed it back – as you say, not a big deal, IntelliJ just happened to mention that it could be a weaker visibility & so I made the warning go away :)

codefromthecrypt commented on zipkin-storage/elasticsearch/src/test/java/zipkin/storage/elasticsearch/ElasticsearchSpanConsumerTest.java:
```diff
@@ -14,6 +14,7 @@
 package zipkin.storage.elasticsearch;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ObjectArrays;
```
imports here are likely a mistake as I don't see any other changes

sethp-jive replied:
Yup, cleaned up now!

codefromthecrypt replied:
ehh.. still here :P

codefromthecrypt commented on zipkin-autoconfigure/storage-elasticsearch/src/main/java/zipkin/autoconfigure/storage/elasticsearch/ZipkinElasticsearchStorageAutoConfiguration.java:
```diff
@@ -13,19 +13,69 @@
  */
 package zipkin.autoconfigure.storage.elasticsearch;
 
+import com.google.common.base.MoreObjects;
```
are all these imports used?

sethp-jive replied:
nope, good catch!

codefromthecrypt commented on zipkin-storage/elasticsearch-jest/README.md:
```diff
@@ -0,0 +1,23 @@
+This module conditionally runs integration tests against a local Elasticsearch instance.
+
+Tests are configured to automatically access Elasticsearch started with its defaults.
+To ensure tests execute, download an Elasticsearch 2.x distribution, extract it, and run `bin/elasticsearch`. 
```
does this work on elasticsearch 5?

sethp-jive replied:
Fairly well – they got rid of the `string` data type (I think we want `keyword`, since we're treating ES as a big `String.equals` engine and not doing "full text" search), but with a few changes I can get many of the tests to pass. The rest fail because ES goes into overdrive and pegs my CPU without returning any results within 10 seconds. The errors it does log suggest I've missed some important fields (i.e. "traceId" is still "text" and not "keyword"), but I stopped there for now.

codefromthecrypt commented on zipkin-storage/elasticsearch-jest/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,398 @@
+
+    @Override
+    public Span deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+        throws JsonParseException {
```
probably don't need throws here and below

sethp-jive replied:
Removed!

codefromthecrypt reviewed (approved):
Looks very good. only some cursory changes.

That said, I'm thinking about re-evaluating the choice I suggested to name this "elasticsearch-jest" This will end up being in server configuration, and I don't expect us to support multiple libraries for elasticsearch http. Similar to how we don't call zipkin-storage-mysql zipkin-storage-jooq, I think we should name this with an "http" suffix afterall.

If I have edit access to this branch, I can change this, if it is a pain..

@anuraaga @shakuzen wdyt?

shakuzen commented on zipkin-autoconfigure/storage-elasticsearch-jest/src/main/java/zipkin/autoconfigure/storage/elasticsearch/jest/ZipkinElasticsearchJestStorageProperties.java:
```diff
@@ -0,0 +1,74 @@
+
+@ConfigurationProperties("zipkin.storage.elasticsearch")
+public class ZipkinElasticsearchJestStorageProperties {
+  /** @see ElasticsearchStorage.Builder#hosts(List) */
```
Per convention, JavaDoc comments on `@ConfigurationProperties` fields are plain text (don't use JavaDoc keywords like `@see`) because they are used by the configuration processor and then by IDEs as descriptions of the fields during auto-completion, for instance.

shakuzen commented on zipkin-storage/elasticsearch-jest/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,398 @@
+    }
+
+    /**
+     * A comma separated list of elasticsearch hostnodes to connect to, in http://host:port or
```
Not a comma-separated list but an actual `List`

sethp-jive committed:
storage: elastic review feedback

Address review feedback, and change the index name to be very likely
unique so as to possibly resolve transient CI failures.

sethp-jive committed:
storage: elasticsearch-jest -> elasticsearch-http

Renaming -jest module to -http.

sethp-jive reviewed (commented):
I made the changes you suggested, as well as fumbled in the dark a bit on the test failures (you'll be unsurprised to hear they pass locally just fine). 

I did go ahead and rename  elasticsearch-jest to elasticsearch-http, but kept it in a separate commit that should be easy to cut out of the history if we want to go back.

codefromthecrypt commented:
This one is my fault I think. So is the best behaviour to copy/paste a
one-liner?

On 20 Sep 2016 12:36 pm, "Tommy Ludwig" notifications@github.com wrote:

> ## _@shakuzen_ commented on this pull request.
> 
> In zipkin-autoconfigure/storage-elasticsearch-jest/src/main/
> java/zipkin/autoconfigure/storage/elasticsearch/jest/
> ZipkinElasticsearchJestStorageProperties.java
> https://github.com/openzipkin/zipkin/pull/1306#pullrequestreview-676674:
> 
> > - \* Unless required by applicable law or agreed to in writing, software distributed under the License
> > - \* is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
> > - \* or implied. See the License for the specific language governing permissions and limitations under
> > - \* the License.
> > - */
> >   +package zipkin.autoconfigure.storage.elasticsearch.jest;
> >   +
> >   +import java.util.Collections;
> >   +import java.util.List;
> >   +import org.springframework.boot.context.properties.ConfigurationProperties;
> >   +import zipkin.storage.elasticsearch.ElasticsearchStorage;
> >   +import zipkin.storage.elasticsearch.jest.RestClient;
> >   +
> >   +@ConfigurationProperties("zipkin.storage.elasticsearch")
> >   +public class ZipkinElasticsearchJestStorageProperties {
> > -  /*\* @see ElasticsearchStorage.Builder#hosts(List) */
> 
> Per convention, JavaDoc comments on @ConfigurationProperties fields are
> plain text (don't use JavaDoc keywords like @see) because they are used
> by the configuration processor and then by IDEs as descriptions of the
> 
> ## fields during auto-completion, for instance.
> 
> In zipkin-storage/elasticsearch-jest/src/main/java/zipkin/
> storage/elasticsearch/jest/RestClient.java
> https://github.com/openzipkin/zipkin/pull/1306#pullrequestreview-676674:
> 
> > -  private static final int MAX_INDICES = 100;
> >   +
> > -  public static final class Builder implements InternalElasticsearchClient.Builder {
> > -    List<String> hosts = Collections.singletonList("http://localhost:9200");
> > -    boolean flushOnWrites;
> >   +
> > -    /**
> > -     \* Unnecessary, but we'll check it's not null for consistency's sake.
> > -     */
> > -    @Override public Builder cluster(String cluster) {
> > -      checkNotNull(cluster, "cluster");
> > -      return this;
> > -    }
> >   +
> > -    /**
> > -     \* A comma separated list of elasticsearch hostnodes to connect to, in http://host:port or
> 
> Not a comma-separated list but an actual List
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1306#pullrequestreview-676674,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAD612Kovuw4gMyVDD5fV63nN5nfusTrks5qr2LXgaJpZM4KBJdk
> .

codefromthecrypt commented on zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,397 @@
+
+    // Create a bulk request when there is more than one span to store
+    ListenableFuture<?> future;
+    final Set<String> indices = new HashSet<>();
```
nit: please use LinkedHashSet/Map always :) order bugs in tests aren't worth tracking down, and such things often happen well after the first commit.

codefromthecrypt commented on zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,397 @@
+        .connTimeout(Ints.checkedCast(TimeUnit.MILLISECONDS.convert(10, TimeUnit.SECONDS)))
+        .readTimeout(Ints.checkedCast(TimeUnit.MILLISECONDS.convert(10, TimeUnit.SECONDS)))
+        .multiThreaded(true)
+        .gson(
```
heh, so jest not only binds to apache hc, but also gson?

sethp-jive replied:
Yeah, but due to repackaging it's a different gson than zipkin core uses.

codefromthecrypt commented on zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,397 @@
+    factory.setHttpClientConfig(new HttpClientConfig.Builder(f.hosts)
+        .defaultMaxTotalConnectionPerRoute(6) // matches "regular" TransportClient node conns
+        .maxTotalConnection(6 * 10) // would be 20 otherwise, or ~3 routes
+        .connTimeout(Ints.checkedCast(TimeUnit.MILLISECONDS.convert(10, TimeUnit.SECONDS)))
```
pretty heavy imho. why not 10 \* 1000? this is fairly obvious milliseconds and the compiler will fold the constant

sethp-jive replied:
I'll change it, but as to the "why" – I would've preferred the java 8 `Duration.ofSeconds(10).toMillis()` (and the checked cast is there because apparently Jest uses ints for time instead of longs), but this is the closest I could come in Java 7.

codefromthecrypt commented on zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  @Override protected void ensureClusterReady(final String catchAll) {
+    class IndicesHealth extends Health {
```
method scoped classes creep me out. they can easily accidentally leak references to the enclosing object. mind making this a static inner class?

codefromthecrypt commented on zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java:
```diff
@@ -0,0 +1,397 @@
+    checkState(!"RED".equalsIgnoreCase(status), "Health status is RED");
+  }
+
+  private Index toIndexRequest(IndexableSpan span) {
```
static utility methods unless they require access to the enclosing object. makes it easier to see impact.

codefromthecrypt commented on zipkin-storage/elasticsearch/src/test/java/zipkin/storage/elasticsearch/ElasticsearchTestGraph.java:
```diff
@@ -28,7 +26,7 @@
       if (ex != null) throw ex;
       ElasticsearchStorage result = ElasticsearchStorage.builder()
-          .index("test_zipkin_native").flushOnWrites(true).build();
+          .index("test_zipkin_" + System.currentTimeMillis()).flushOnWrites(true).build();
```
please revert this. I don't want my ES polluted

codefromthecrypt reviewed (commented):
found a few other things on second look

sethp-jive committed:
storage: more review feedback

Address additional review feedback, including reverting the change to
 index naming.

codefromthecrypt commented on zipkin-autoconfigure/storage-elasticsearch-http/src/main/java/zipkin/autoconfigure/storage/elasticsearch/http/ZipkinElasticsearchHttpStorageProperties.java:
```diff
@@ -0,0 +1,76 @@
+public class ZipkinElasticsearchHttpStorageProperties {
+  /** The elasticsearch cluster to connect to, defaults to "elasticsearch". */
+  private String cluster = "elasticsearch";
+  /** A List of hosts to connect to, e.g.  */
```
"A List of base http urls to connect to, e.g. http://localhost:9200" right?

codefromthecrypt commented on zipkin-autoconfigure/storage-elasticsearch-http/src/main/java/zipkin/autoconfigure/storage/elasticsearch/http/ZipkinElasticsearchHttpStorageProperties.java:
```diff
@@ -0,0 +1,76 @@
+  /** The elasticsearch cluster to connect to, defaults to "elasticsearch". */
+  private String cluster = "elasticsearch";
+  /** A List of hosts to connect to, e.g.  */
+  private List<String> hosts = Collections.singletonList("localhost:9300");
```
http://localhost:9200, right?

codefromthecrypt commented on zipkin-autoconfigure/storage-elasticsearch/src/main/java/zipkin/autoconfigure/storage/elasticsearch/ZipkinElasticsearchStorageAutoConfiguration.java:
```diff
@@ -13,19 +13,65 @@
+    private static boolean hostsAreUrls(String hostsProperty) {
+      List<String> hosts = Arrays.asList(hostsProperty.split(","));
+      try {
+        URL url = new URL(Iterables.getFirst(hosts, ""));
```
:) I used to over-use guava like this.. it is hard to realize when it is happening.

codefromthecrypt reviewed (approved):
I'll do some last tidying and squash the interim commits on the way in (leaving the separate one for http as you mentioned)

great work

codefromthecrypt commented:
cherry-picked in. take note of a few conventional file renames (also it is now plumbed to zipkin-server), so the following works:

``` bash
$ STORAGE_TYPE=elasticsearch ES_HOSTS=http://localhost:9200 java -jar zipkin-server/target/zipkin-server-1.11.2-SNAPSHOT-exec.jar
```
